optuna/integration/__init__.py

```diff
@@ -1,6 +1,65 @@
-from optuna.integration.chainer import ChainerPruningExtension  # NOQA
-from optuna.integration.chainermn import ChainerMNStudy  # NOQA
-from optuna.integration.keras import KerasPruningCallback  # NOQA
-from optuna.integration.lightgbm import LightGBMPruningCallback  # NOQA
-from optuna.integration.tensorflow import TensorFlowPruningHook  # NOQA
-from optuna.integration.xgboost import XGBoostPruningCallback  # NOQA
+import os
+import sys
+from types import ModuleType
+from typing import Any  # NOQA
+
+from optuna.types import TYPE_CHECKING
+
+
+_import_structure = {
+    'chainer': ['ChainerPruningExtension'],
+    'chainermn': ['ChainerMNStudy'],
+    'keras': ['KerasPruningCallback'],
+    'lightgbm': ['LightGBMPruningCallback'],
+    'tensorflow': ['TensorFlowPruningHook'],
+    'xgboost': ['XGBoostPruningCallback'],
+}
+
+
+__all__ = list(_import_structure.keys()) + sum(_import_structure.values(), [])
```

Comment:
Do we want to expose submodules such as `optuna.integration.chainer` in addition to classes such as `optuna.integration.ChainerPruningExtension` ?

I think it is totally reasonable to expose them. For example, we can keep backward compatibility (although I think there are few users who are importing these sub modules).

On the other hand, some of famous Python packages such as scikit-learn do not expose unnecessary submodules by using `__all__` everywhere (e.g., https://github.com/scikit-learn/scikit-learn/blob/a7e17117bb15eb3f51ebccc1bd53e42fcb4e6cd8/sklearn/mixture/__init__.py), and I thought this may also be possible future choice. I think we can change the policy in future, so I think we do not need to decide now, and I do not suggest to change here now.

Reply:
As you pointed out, we've better discuss in the future if exposing submodules actually makes sense.

For now, let's keep it as is for backward compatibility.